torchvision/datasets/utils.py

```diff
@@ -402,16 +417,24 @@ def extract_archive(from_path: str, to_path: Optional[str] = None, remove_finish
     Returns:
         (str): Path to the directory the file was extracted to.
     """
+
+    def path_or_str(ret_path: pathlib.Path) -> Union[str, pathlib.Path]:
+        if isinstance(from_path, str):
+            return os.fspath(ret_path)
+        else:
+            return ret_path
+
     if to_path is None:
         to_path = os.path.dirname(from_path)
 
     suffix, archive_type, compression = _detect_file_type(from_path)
     if not archive_type:
-        return _decompress(
+        ret_path = _decompress(
             from_path,
             os.path.join(to_path, os.path.basename(from_path).replace(suffix, "")),
             remove_finished=remove_finished,
         )
+        return path_or_str(ret_path)
```

Comment:
Would it make more sense to call `path_or_str` within `_decompress` so that only one of these functions gets changed?

Reply:
One issue is that that the os.path.dirname on L406 converts pathlib.Path into str, so to_path is always str if the code goes through that if condition:

```
>>> import pathlib
>>> import os
>>> a_str="/a/b"
>>> a_path=pathlib.Path(a_str)
>>> a_str
'/a/b'
>>> a_path
PosixPath('/a/b')
>>> os.path.dirname(a_str)
'/a'
>>> os.path.dirname(a_path)
'/a'

```

So if `archive_type` is not None, we can't just return `to_path` without checking the type of `from_path`.

Your thoughts?

Reply:
OK good points, let's keep it as-is then